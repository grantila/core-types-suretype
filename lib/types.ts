import type { WarnFunction } from 'core-types'
import type { ExportRefMethod } from 'suretype'
import type { Options as AjvOptions } from 'ajv'

type PickByType<T, Value> = {
	[P in keyof T as T[P] extends Value | undefined ? P : never]: T[P]
}

export interface JsonSchemaToSuretypeOptions
{
	warn?: WarnFunction;

	/**
	 * The name of the file the result will be written to
	 */
	filename?: string;

	/**
	 * The name of the source file used to convert to TypeScript
	 */
	sourceFilename?: string;

	/**
	 * Use `unknown` instead of `any` for any-types.
	 *
	 * @default true
	 */
	useUnknown?: boolean;

	/**
	 * The name of the package using this package to perform the conversion
	 */
	userPackage?: string;

	/**
	 * The url to the package using this package
	 */
	userPackageUrl?: string;

	/**
	 * Set to true to prevent writing the "disable linting" comment
	 */
	noDisableLintHeader?: boolean;

	/**
	 * Set to true to not write a top-level descriptive comment about the
	 * auto-generated file
	 */
	noDescriptiveHeader?: boolean;


	/**
	 * Forward the JSON Schema, and create an untyped validator schema with
	 * the raw JSON Schema under the hood.
	 *
	 * Typings will still be created, and the validators will coerce to them.
	 *
	 * @default false
	 */
	forwardSchema?: boolean;

	/**
	 * Inline pretty typescript types aside validator code
	 *
	 * This defaults to true, and is forced to true if forwardSchema is true
	 *
	 * @default true
	 */
	inlineTypes?: boolean;

	/**
	 * Export the deduced types (or the pretty types, depending on inlineTypes)
	 *
	 * @default true
	 */
	exportType?: boolean;

	/**
	 * Export validator schemas
	 *
	 * @default true
	 */
	exportSchema?: boolean;

	/**
	 * Export regular validators
	 *
	 * @default true
	 */
	exportValidator?: boolean;

	/**
	 * Export 'ensurer' validators
	 *
	 * @default true
	 */
	exportEnsurer?: boolean;

	/**
	 * Export type guards (is* validators)
	 *
	 * @default true
	 */
	exportTypeGuard?: boolean;

	/**
	 * Subset of AJV options that can be set for all compile calls
	 */
	ajvOptions?: PickByType<AjvOptions, boolean | string>

	/**
	 * What to do when detecting an unsupported type
	 *
	 *  - `ignore`: Ignore; skip type or cast to any
	 *  - `warn`: Same as 'ignore', but warn (default)
	 *  - `error`: Throw an error
	 */
	unsupported?: 'ignore' | 'warn' | 'error';


	/**
	 * What to do when detecting an unresolvable reference
	 *
	 *  - `ignore`: Ignore; skip type or cast to any
	 *  - `warn`: Same as 'ignore', but warn (default)
	 *  - `error`: Throw an error
	 */
	missingReference?: 'ignore' | 'warn' | 'error';
}

export interface SuretypeToJsonSchemaOptions
{
	warn?: WarnFunction;


	/**
	 * The name of the file the result will be written to
	 */
	filename?: string;

	/**
	 * The name of the source file used to convert to TypeScript
	 */
	sourceFilename?: string;

	/**
	 * The name of the package using this package to perform the conversion
	 */
	userPackage?: string;

	/**
	 * The url to the package using this package
	 */
	userPackageUrl?: string;

	refMethod?: ExportRefMethod;

	/**
	 * What to do when detecting name conflicts
	 *
	 *  - `rename`: Try to rename type
	 *  - `warn`: Ignore type, but warn
	 *  - `error`: Throw an error (default)
	 */
	nameConflict?: 'rename' | 'warn' | 'error';
}
