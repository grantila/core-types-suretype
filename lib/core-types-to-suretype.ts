import type { ConversionResult, NodeDocument } from 'core-types'
import { convertCoreTypesToJsonSchema } from 'core-types-json-schema'

import { convertJsonSchemaToSureType } from './json-schema-to-suretype.js'
import type { JsonSchemaToSuretypeOptions } from './types.js'


export function convertCoreTypesToSureType(
	doc: NodeDocument,
	options: JsonSchemaToSuretypeOptions
)
: ConversionResult< string >
{
	const resJsonSchema = convertCoreTypesToJsonSchema( doc, options );
	const { data: jsonSchema } = resJsonSchema;

	const resSureType = convertJsonSchemaToSureType( jsonSchema, options );

	return {
		convertedTypes: resSureType.convertedTypes,
		notConvertedTypes: [ ...new Set( [
			...resJsonSchema.notConvertedTypes,
			...resSureType.notConvertedTypes,
		] ) ],
		data: resSureType.data,
	};
}
