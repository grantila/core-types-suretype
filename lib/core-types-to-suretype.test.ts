import { NamedType, NodeDocument } from 'core-types'

import { convertCoreTypesToSureType } from './core-types-to-suretype.js'
import { JsonSchemaToSuretypeOptions } from './types.js'


const wrapDocument = ( types: Array< NamedType > ): NodeDocument =>
	( {
		version: 1,
		types
	} );

const options: JsonSchemaToSuretypeOptions = {
	useUnknown: true,
	inlineTypes: true,
	exportSchema: true,
	exportType: true,
	exportValidator: true,
	exportEnsurer: true,
	exportTypeGuard: true,
	forwardSchema: false,
};

describe( 'core-types-to-suretype', ( ) =>
{
	it( 'simple type', ( ) =>
	{
		const { convertedTypes, notConvertedTypes, data: ts } =
			convertCoreTypesToSureType( wrapDocument( [
				{
					name: 'foo',
					type: 'string',
				}
			] ), options );
		expect( convertedTypes ).toStrictEqual( [ 'foo' ] );
		expect( notConvertedTypes ).toStrictEqual( [ ] );
		expect( ts ).toMatchSnapshot( );
	} );

	it( 'simple string union of separate types', ( ) =>
	{
		const { convertedTypes, notConvertedTypes, data: ts } =
			convertCoreTypesToSureType( wrapDocument( [
				{
					name: 'bar',
					type: 'string',
					const: 'bar',
				},
				{
					name: 'foo',
					type: 'or',
					or: [
						{
							type: 'ref',
							ref: 'bar',
						},
						{
							type: 'string',
							enum: [ 'foo', 'baz' ],
						}
					]
				}
			] ), options );
		expect( convertedTypes ).toStrictEqual( [ 'bar', 'foo' ] );
		expect( notConvertedTypes ).toStrictEqual( [ ] );
		expect( ts ).toMatchSnapshot( );
	} );

	it( 'complex type', ( ) =>
	{
		const res = convertCoreTypesToSureType( wrapDocument( [
			{
				name: 'User',
				type: 'string'
			},
			{
				name: 'foo',
				type: 'object',
				properties: {
					bar: { required: false, node: { type: 'string' } },
					baz: {
						required: true,
						node: {
							type: 'or',
							or: [
								{ type: 'number' },
								{
									type: 'object',
									properties: { },
									additionalProperties: { type: 'boolean' },
								}
							]
						},
					},
					bak: {
						required: true,
						node: {
							type: 'object',
							properties: { },
							additionalProperties: true,
						},
					},
					tupleWithAdditionals: {
						required: true,
						node: {
							type: 'tuple',
							elementTypes: [ { type: 'string' } ],
							additionalItems: { type: 'number' },
							minItems: 1,
						},
					},
					tupleWithObjectAdditionals: {
						required: true,
						node: {
							type: 'tuple',
							elementTypes: [ { type: 'string' } ],
							additionalItems: {
								type: 'object',
								properties: {
									x: {
										required: false,
										node: { type: 'ref', ref: 'User' },
									},
								},
								additionalProperties: false,
							},
							minItems: 1,
						},
					},
				},
				additionalProperties: false,
			}
		] ),
		options
		);
		const { convertedTypes, notConvertedTypes, data: ts } = res;
		expect( convertedTypes ).toStrictEqual( [ 'User', 'foo' ] );
		expect( notConvertedTypes ).toStrictEqual( [ ] );
		expect( ts ).toMatchSnapshot( );
	} );

	it( 'used in readme', ( ) =>
	{
		const res = convertCoreTypesToSureType( wrapDocument( [
			{
				name: 'User',
				title: 'User',
				description:
					'User type\n\n' +
					'This type holds the user information, such as name',
				type: 'object',
				properties: {
					name: {
						node: { type: 'string', description: 'The real name' },
						required: true
					},
				},
				additionalProperties: false,
			},
			{
				name: 'ChatLine',
				title: 'ChatLine',
				description: 'A chat line',
				type: 'object',
				properties: {
					user: {
						node: { type: 'ref', ref: 'User' }, required: true
					},
					line: { node: { type: 'string' }, required: true },
				},
				additionalProperties: false,
			},
		] ), options );

		const { convertedTypes, notConvertedTypes, data: ts } = res;

		expect( convertedTypes ).toStrictEqual( [ 'User', 'ChatLine' ] );
		expect( notConvertedTypes ).toStrictEqual( [ ] );
		expect( ts ).toMatchSnapshot( );
	} );

	it( 'should write annotations properly', ( ) =>
	{
		const res = convertCoreTypesToSureType( wrapDocument( [
			{
				name: 'User',
				title: 'User',
				description:
					'User type\n\n' +
					'This type holds the user information, such as name',
				examples: [ '{ name: "Joe" }' ],
				default: '{ user: "" }',
				see: [ 'http://username' ],
				type: 'object',
				properties: {
					name: {
						node: {
							type: 'string',
							title: 'User.name',
							description:
								'The real name\n\n' +
								'Must be a valid name, not */'
						},
						required: true
					},
				},
				additionalProperties: false,
			},
			{
				name: 'ChatLine',
				title: 'ChatLine',
				description: 'A chat line',
				type: 'object',
				properties: {
					user: {
						node: {
							title: 'ChatLine.user',
							description: 'User ref',
							type: 'ref',
							ref: 'User'
						},
						required: true,
					},
					line: {
						node: {
							examples: 'This is a line',
							type: 'string',
						},
						required: true
					},
				},
				additionalProperties: false,
			},
			{
				name: 'Thing',
				type: 'any',
			},
			{
				title: 'Thingy',
				description: 'Thing ref',
				name: 'Thingy',
				type: 'or',
				or: [
					{
						type: 'ref',
						ref: 'Thing',
						title: 'Thingy',
						description: 'Thing is the preferred type',
						see: 'The Thing documentation',
					},
					{
						type: 'number',
						title: 'Thingy',
						description: 'Just a number',
					},
				],
			},
		] ), options );

		const { convertedTypes, notConvertedTypes, data: ts } = res;

		expect( convertedTypes.sort( ) )
			.toStrictEqual( [ 'ChatLine', 'Thing', 'Thingy', 'User' ] );
		expect( notConvertedTypes ).toStrictEqual( [ ] );
		expect( ts ).toMatchSnapshot( );
	} );
} );
