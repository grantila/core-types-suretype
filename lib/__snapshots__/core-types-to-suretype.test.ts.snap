// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core-types-to-suretype complex type 1`] = `
"/* tslint:disable */
/* eslint-disable */
/**
 * This file is generated by core-types-suretype, DO NOT EDIT.
 * For more information, see:
 *  - {@link https://github.com/grantila/core-types-suretype}
 */

import { suretype, v, compile } from 'suretype';

/** The validation schema for a User */
export const schemaUser = suretype({
    name: "User"
}, v.string());

export type User = string;

/**
 * ## Validate that a variable is a User
 *
 * @returns ValidationResult
 */
export const validateUser = compile(schemaUser);

/**
 * ## Ensure a variable is a User
 *
 * This call will throw a ValidationError if the variable isn't a User.
 *
 * If the variable **is** a User, the returned variable will be of that type.
 */
export const ensureUser = compile<typeof schemaUser, User>(schemaUser, { ensure: true });

/**
 * ## Validates that a variable is a User
 *
 * @returns boolean
 */
export const isUser = compile(schemaUser, { simple: true });

/** The validation schema for a foo */
export const schemaFoo = suretype({
    name: "foo"
}, v.object({
    bar: v.string(),
    baz: v.anyOf([v.number(), v.object({}).additional(v.boolean())]).required(),
    bak: v.object({}).additional(true).required(),
    tupleWithAdditionals: v.array([
        v.string()
    ]).minItems(1).additional(v.number()).required(),
    tupleWithObjectAdditionals: v.array([
        v.string()
    ]).minItems(1).additional(v.object({
        x: schemaUser
    })).required()
}));

export interface foo {
    bar?: string;
    baz: number | {
        [key: string]: boolean;
    };
    bak: {
        [key: string]: unknown;
    };
    tupleWithAdditionals: [
        string,
        ...number[]
    ];
    tupleWithObjectAdditionals: [
        string,
        ...{
            x?: User;
        }[]
    ];
}

/**
 * ## Validate that a variable is a foo
 *
 * @returns ValidationResult
 */
export const validateFoo = compile(schemaFoo);

/**
 * ## Ensure a variable is a foo
 *
 * This call will throw a ValidationError if the variable isn't a foo.
 *
 * If the variable **is** a foo, the returned variable will be of that type.
 */
export const ensureFoo = compile<typeof schemaFoo, foo>(schemaFoo, { ensure: true });

/**
 * ## Validates that a variable is a foo
 *
 * @returns boolean
 */
export const isFoo = compile(schemaFoo, { simple: true });
"
`;

exports[`core-types-to-suretype should write annotations properly 1`] = `
"/* tslint:disable */
/* eslint-disable */
/**
 * This file is generated by core-types-suretype, DO NOT EDIT.
 * For more information, see:
 *  - {@link https://github.com/grantila/core-types-suretype}
 */

import { suretype, v, compile, annotate } from 'suretype';

/** The validation schema for a Thing */
export const schemaThing = suretype({
    name: "Thing"
}, v.unknown());

export type Thing = unknown;

/**
 * ## Validate that a variable is a Thing
 *
 * @returns ValidationResult
 */
export const validateThing = compile(schemaThing);

/**
 * ## Ensure a variable is a Thing
 *
 * This call will throw a ValidationError if the variable isn't a Thing.
 *
 * If the variable **is** a Thing, the returned variable will be of that type.
 */
export const ensureThing = compile<typeof schemaThing, Thing>(schemaThing, { ensure: true });

/**
 * ## Validates that a variable is a Thing
 *
 * @returns boolean
 */
export const isThing = compile(schemaThing, { simple: true });

/** The validation schema for a Thingy */
export const schemaThingy = suretype({
    name: "Thingy",
    title: "Thingy",
    description: "Thing ref"
}, v.anyOf([annotate({
        title: "Thingy",
        description: "Thing is the preferred type\\n\\n@see The Thing documentation"
    }, schemaThing), annotate({
        title: "Thingy",
        description: "Just a number"
    }, v.number())]));

export type Thingy = Thing | /** Just a number */
number;

/**
 * ## Validate that a variable is a Thingy
 *
 * @returns ValidationResult
 */
export const validateThingy = compile(schemaThingy);

/**
 * ## Ensure a variable is a Thingy
 *
 * This call will throw a ValidationError if the variable isn't a Thingy.
 *
 * If the variable **is** a Thingy, the returned variable will be of that type.
 */
export const ensureThingy = compile<typeof schemaThingy, Thingy>(schemaThingy, { ensure: true });

/**
 * ## Validates that a variable is a Thingy
 *
 * @returns boolean
 */
export const isThingy = compile(schemaThingy, { simple: true });

/** The validation schema for a User */
export const schemaUser = suretype({
    name: "User",
    title: "User",
    examples: ["[\\n  \\"{ name: \\\\\\"Joe\\\\\\" }\\"\\n]"],
    description: "User type\\n\\nThis type holds the user information, such as name\\n\\n@see http://username\\n\\n@default\\n    { user: \\"\\" }"
}, v.object({
    name: annotate({
        title: "User.name",
        description: "The real name\\n\\nMust be a valid name, not */"
    }, v.string().required())
}).default("{ user: \\"\\" }"));

/**
 * User type
 *
 * This type holds the user information, such as name
 *
 * @example
 *     [
 *       "{ name: \\"Joe\\" }"
 *     ]
 *
 * @default
 *     { user: "" }
 *
 * @see http://username
 */
export interface User {
    /**
     * The real name
     *
     * Must be a valid name, not *\\/
     */
    name: string;
}

/**
 * ## Validate that a variable is a User
 *
 * @returns ValidationResult
 */
export const validateUser = compile(schemaUser);

/**
 * ## Ensure a variable is a User
 *
 * This call will throw a ValidationError if the variable isn't a User.
 *
 * If the variable **is** a User, the returned variable will be of that type.
 */
export const ensureUser = compile<typeof schemaUser, User>(schemaUser, { ensure: true });

/**
 * ## Validates that a variable is a User
 *
 * @returns boolean
 */
export const isUser = compile(schemaUser, { simple: true });

/** The validation schema for a ChatLine */
export const schemaChatLine = suretype({
    name: "ChatLine",
    title: "ChatLine",
    description: "A chat line"
}, v.object({
    user: annotate({
        title: "ChatLine.user",
        description: "User ref"
    }, schemaUser),
    line: annotate({
        examples: ["This is a line"]
    }, v.string().required())
}));

/** A chat line */
export interface ChatLine {
    user: User;
    /**
     * @example
     *     This is a line
     */
    line: string;
}

/**
 * ## Validate that a variable is a ChatLine
 *
 * @returns ValidationResult
 */
export const validateChatLine = compile(schemaChatLine);

/**
 * ## Ensure a variable is a ChatLine
 *
 * This call will throw a ValidationError if the variable isn't a ChatLine.
 *
 * If the variable **is** a ChatLine, the returned variable will be of that type.
 */
export const ensureChatLine = compile<typeof schemaChatLine, ChatLine>(schemaChatLine, { ensure: true });

/**
 * ## Validates that a variable is a ChatLine
 *
 * @returns boolean
 */
export const isChatLine = compile(schemaChatLine, { simple: true });
"
`;

exports[`core-types-to-suretype simple string union of separate types 1`] = `
"/* tslint:disable */
/* eslint-disable */
/**
 * This file is generated by core-types-suretype, DO NOT EDIT.
 * For more information, see:
 *  - {@link https://github.com/grantila/core-types-suretype}
 */

import { suretype, v, compile } from 'suretype';

/** The validation schema for a bar */
export const schemaBar = suretype({
    name: "bar"
}, v.string().const("bar"));

export type bar = "bar";

/**
 * ## Validate that a variable is a bar
 *
 * @returns ValidationResult
 */
export const validateBar = compile(schemaBar);

/**
 * ## Ensure a variable is a bar
 *
 * This call will throw a ValidationError if the variable isn't a bar.
 *
 * If the variable **is** a bar, the returned variable will be of that type.
 */
export const ensureBar = compile<typeof schemaBar, bar>(schemaBar, { ensure: true });

/**
 * ## Validates that a variable is a bar
 *
 * @returns boolean
 */
export const isBar = compile(schemaBar, { simple: true });

/** The validation schema for a foo */
export const schemaFoo = suretype({
    name: "foo"
}, v.anyOf([schemaBar, v.string().enum("foo", "baz")]));

export type foo = bar | ("foo" | "baz");

/**
 * ## Validate that a variable is a foo
 *
 * @returns ValidationResult
 */
export const validateFoo = compile(schemaFoo);

/**
 * ## Ensure a variable is a foo
 *
 * This call will throw a ValidationError if the variable isn't a foo.
 *
 * If the variable **is** a foo, the returned variable will be of that type.
 */
export const ensureFoo = compile<typeof schemaFoo, foo>(schemaFoo, { ensure: true });

/**
 * ## Validates that a variable is a foo
 *
 * @returns boolean
 */
export const isFoo = compile(schemaFoo, { simple: true });
"
`;

exports[`core-types-to-suretype simple type 1`] = `
"/* tslint:disable */
/* eslint-disable */
/**
 * This file is generated by core-types-suretype, DO NOT EDIT.
 * For more information, see:
 *  - {@link https://github.com/grantila/core-types-suretype}
 */

import { suretype, v, compile } from 'suretype';

/** The validation schema for a foo */
export const schemaFoo = suretype({
    name: "foo"
}, v.string());

export type foo = string;

/**
 * ## Validate that a variable is a foo
 *
 * @returns ValidationResult
 */
export const validateFoo = compile(schemaFoo);

/**
 * ## Ensure a variable is a foo
 *
 * This call will throw a ValidationError if the variable isn't a foo.
 *
 * If the variable **is** a foo, the returned variable will be of that type.
 */
export const ensureFoo = compile<typeof schemaFoo, foo>(schemaFoo, { ensure: true });

/**
 * ## Validates that a variable is a foo
 *
 * @returns boolean
 */
export const isFoo = compile(schemaFoo, { simple: true });
"
`;

exports[`core-types-to-suretype used in readme 1`] = `
"/* tslint:disable */
/* eslint-disable */
/**
 * This file is generated by core-types-suretype, DO NOT EDIT.
 * For more information, see:
 *  - {@link https://github.com/grantila/core-types-suretype}
 */

import { suretype, v, compile, annotate } from 'suretype';

/** The validation schema for a User */
export const schemaUser = suretype({
    name: "User",
    title: "User",
    description: "User type\\n\\nThis type holds the user information, such as name"
}, v.object({
    name: annotate({
        description: "The real name"
    }, v.string().required())
}));

/**
 * User type
 *
 * This type holds the user information, such as name
 */
export interface User {
    /** The real name */
    name: string;
}

/**
 * ## Validate that a variable is a User
 *
 * @returns ValidationResult
 */
export const validateUser = compile(schemaUser);

/**
 * ## Ensure a variable is a User
 *
 * This call will throw a ValidationError if the variable isn't a User.
 *
 * If the variable **is** a User, the returned variable will be of that type.
 */
export const ensureUser = compile<typeof schemaUser, User>(schemaUser, { ensure: true });

/**
 * ## Validates that a variable is a User
 *
 * @returns boolean
 */
export const isUser = compile(schemaUser, { simple: true });

/** The validation schema for a ChatLine */
export const schemaChatLine = suretype({
    name: "ChatLine",
    title: "ChatLine",
    description: "A chat line"
}, v.object({
    user: schemaUser,
    line: v.string().required()
}));

/** A chat line */
export interface ChatLine {
    user: User;
    line: string;
}

/**
 * ## Validate that a variable is a ChatLine
 *
 * @returns ValidationResult
 */
export const validateChatLine = compile(schemaChatLine);

/**
 * ## Ensure a variable is a ChatLine
 *
 * This call will throw a ValidationError if the variable isn't a ChatLine.
 *
 * If the variable **is** a ChatLine, the returned variable will be of that type.
 */
export const ensureChatLine = compile<typeof schemaChatLine, ChatLine>(schemaChatLine, { ensure: true });

/**
 * ## Validates that a variable is a ChatLine
 *
 * @returns boolean
 */
export const isChatLine = compile(schemaChatLine, { simple: true });
"
`;
